### PR TITLE
Offload file operations to prevent the event loop from being blocked by synchronous file operations

### LINKED ISSUE #2520: [Workaround Patch Available] Blocking call in aiohttp CookieJar causing stability issues
**Describe the bug**

The `alexa_media` integration is causing blocking calls within the event loop due to synchronous file operations in `aiohttp.cookiejar.CookieJar`. Specifically, the methods `save` and `load` in `CookieJar` perform synchronous file I/O operations, leading to stability issues in Home Assistant and generating warnings about blocking calls to `open`.

**To Reproduce**

1. Install and configure the `alexa_media` integration in Home Assistant.
2. Start Home Assistant.
3. Monitor the logs for warnings or errors.
4. The following error messages appear, indicating blocking calls:

```
Detected blocking call to open with args (PosixPath('/config/.storage/alexa_media.YOUR_EMAIL@gmail.com.pickle'),) in /usr/local/lib/python3.12/site-packages/aiohttp/cookiejar.py, line 113: with file_path.open(mode="wb") as f: inside the event loop; This is causing stability issues. Please create a bug report at https://github.com/home-assistant/core/issues?q=is%3Aopen+is%3Aissue For developers, please see https://developers.home-assistant.io/docs/asyncio_blocking_operations/#open
```

And:

```
Detected blocking call to open with args (PosixPath('/config/.storage/alexa_media.YOUR_EMAIL@gmail.com.pickle'),) inside the event loop by custom integration 'alexa_media' at custom_components/alexa_media/__init__.py, line 251: cookies = await login.load_cookie() (offender: /usr/local/lib/python3.12/site-packages/aiohttp/cookiejar.py, line 118: with file_path.open(mode="rb") as f:), please create a bug report at https://github.com/alandtse/alexa_media_player/issues For developers, please see https://developers.home-assistant.io/docs/asyncio_blocking_operations/#open
```

**Expected behavior**

Asynchronous operations should not perform blocking file I/O within the event loop. The integration should avoid blocking calls to maintain stability in Home Assistant.

**Screenshots**

N/A

**System details**

- Home Assistant version: 2024.9.1
- `alexa_media` version (from `const.py` or HA startup log): e.g., 4.12.12
- `alexapy` version (from `pip show alexapy` in Home Assistant container or HA startup log): e.g., 1.29.2
- Is Amazon [2FA/2SV enabled](https://github.com/alandtse/alexa_media_player/wiki/Configuration#enable-two-factor-authentication-for-your-amazon-account) (y/n): Yes
- Amazon Domain: amazon.de

**Debug Logs (alexa_media & alexapy)**

```plaintext
Detected blocking call to open with args (PosixPath('/config/.storage/alexa_media.YOUR_EMAIL@gmail.com.pickle'),) in /usr/local/lib/python3.12/site-packages/aiohttp/cookiejar.py, line 113: with file_path.open(mode="wb") as f: inside the event loop; This is causing stability issues. Please create a bug report at https://github.com/home-assistant/core/issues?q=is%3Aopen+is%3Aissue For developers, please see https://developers.home-assistant.io/docs/asyncio_blocking_operations/#open

Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "/usr/src/homeassistant/homeassistant/__main__.py", line 223, in <module>
    sys.exit(main())
  File "/usr/src/homeassistant/homeassistant/__main__.py", line 209, in main
    exit_code = runner.run(runtime_conf)
  File "/usr/src/homeassistant/homeassistant/runner.py", line 189, in run
    return loop.run_until_complete(setup_and_run_hass(runtime_config))
  File "/usr/local/lib/python3.12/asyncio/base_events.py", line 674, in run_until_complete
    self.run_forever()
  File "/usr/local/lib/python3.12/asyncio/base_events.py", line 641, in run_forever
    self._run_once()
  File "/usr/local/lib/python3.12/asyncio/base_events.py", line 1990, in _run_once
    handle._run()
  File "/usr/local/lib/python3.12/asyncio/events.py", line 88, in _run
    self._context.run(self._callback, *self._args)
  File "/usr/local/lib/python3.12/site-packages/alexapy/helpers.py", line 137, in wrapper
    return await func(*args, **kwargs)
  File "/usr/local/lib/python3.12/site-packages/alexapy/alexaapi.py", line 1228, in get_devices
    response = await AlexaAPI._static_request(
  File "/usr/local/lib/python3.12/site-packages/backoff/_async.py", line 151, in retry
    ret = await target(*args, **kwargs)
  File "/usr/local/lib/python3.12/site-packages/alexapy/alexaapi.py", line 281, in _static_request
    await login.save_cookiefile()
  File "/usr/local/lib/python3.12/site-packages/alexapy/alexalogin.py", line 719, in save_cookiefile
    self._outputpath(f"{self._hass_domain}.{self.email}.pickle"),
  File "/usr/local/lib/python3.12/site-packages/aiohttp/cookiejar.py", line 113, in save
    with file_path.open(mode="wb") as f:
```

Additional similar error during cookie loading:

```plaintext
Detected blocking call to open with args (PosixPath('/config/.storage/alexa_media.YOUR_EMAIL@gmail.com.pickle'),) inside the event loop by custom integration 'alexa_media' at custom_components/alexa_media/__init__.py, line 251: cookies = await login.load_cookie() (offender: /usr/local/lib/python3.12/site-packages/aiohttp/cookiejar.py, line 118: with file_path.open(mode="rb") as f:), please create a bug report at https://github.com/alandtse/alexa_media_player/issues For developers, please see https://developers.home-assistant.io/docs/asyncio_blocking_operations/#open

Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "/usr/src/homeassistant/homeassistant/__main__.py", line 223, in <module>
    sys.exit(main())
  File "/usr/src/homeassistant/homeassistant/__main__.py", line 209, in main
    exit_code = runner.run(runtime_conf)
  File "/usr/src/homeassistant/homeassistant/runner.py", line 189, in run
    return loop.run_until_complete(setup_and_run_hass(runtime_config))
  File "/usr/local/lib/python3.12/asyncio/base_events.py", line 674, in run_until_complete
    self.run_forever()
  File "/usr/local/lib/python3.12/asyncio/base_events.py", line 641, in run_forever
    self._run_once()
  File "/usr/local/lib/python3.12/asyncio/base_events.py", line 1990, in _run_once
    handle._run()
  File "/usr/local/lib/python3.12/asyncio/events.py", line 88, in _run
    self._context.run(self._callback, *self._args)
  File "/usr/src/homeassistant/homeassistant/config_entries.py", line 752, in async_setup_locked
    await self.async_setup(hass, integration=integration)
  File "/usr/src/homeassistant/homeassistant/config_entries.py", line 604, in async_setup
    result = await component.async_setup_entry(hass, self)
  File "/config/custom_components/alexa_media/__init__.py", line 474, in async_setup_entry
    cookies = await perform_login(login)
  File "/config/custom_components/alexa_media/__init__.py", line 251, in perform_login
    cookies = await login.load_cookie()
```

**Additional context**

Please find the adjusted [`cookiejar.py`](https://github.com/user-attachments/files/16999127/cookiejar-patched.py.txt) file attached or consider integrating similar changes to address the issue.

You can apply this patch by:

1. **Install the [Patch integration](https://github.com/amitfin/patch) with HACS.**

2. **Download** the `cookiejar-patched.py.txt` file, rename it to `cookiejar.py`, and place it in a `config/patches/aiohttp/3.10.5/patch` directory.

3. **Copy** the original `/usr/local/lib/python3.12/site-packages/aiohttp/cookiejar.py` to `config/patches/aiohttp/3.10.5/base`.

Then add this to your `configuration.yaml`:

```yaml
patch:
  delay: 10
  restart: true
  files:
    - name: cookiejar.py
      base: patches/aiohttp/3.10.5/base
      destination: "/usr/local/lib/python3.12/site-packages/aiohttp"
      patch: patches/aiohttp/3.10.5/patch
```

This workaround modifies the `cookiejar.py` file to offload blocking file operations to a separate thread and ensures synchronization using `threading.Event`. This allows the methods to remain synchronous from the caller's perspective, avoiding changes to `alexapy` or `alexa_media`.

---

**Proposed Solution**

Implement the adjusted `cookiejar.py` to resolve the blocking call issue. The key changes include:

- **Offloading file operations** in the `save` and `load` methods to separate threads using `run_in_executor`.
- **Using `threading.Event`** to wait for the completion of loading and saving operations.
- **Ensuring synchronization**, so methods accessing `self._cookies` wait until the cookies are fully loaded.

This adjustment prevents the event loop from being blocked by synchronous file operations and maintains compatibility with existing code that relies on `aiohttp.cookiejar.CookieJar`.

---

**Note:** While this workaround resolves the immediate problem, a long-term solution would involve updating `aiohttp` to handle cookie file operations asynchronously or adjusting `alexapy` and `alexa_media` to avoid blocking calls within the event loop. 

Additionally, if this workaround proves effective, it would be beneficial for someone to open an issue on the [aiohttp GitHub repository](https://github.com/aio-libs/aiohttp/issues) to address this problem.

### PR DESCRIPTION
**Pull request by danielbrunt57**
## What do these changes do?

- Offload file operations in the save and load methods to separate threads using run_in_executor.
- Use threading.Event to wait for the completion of loading and saving operations.
- Ensure synchronization, so methods accessing self._cookies wait until the cookies are fully loaded.

## Are there changes in behavior for the user?

No notable behaviour change for end users.

## Is it a substantial burden for the maintainers to support this?

Minimal to no burden is expected for maintainers to support these changes but will eliminate these warnings in Home Assistant:
> Detected blocking call to open with args (PosixPath('/config/.storage/alexa_media.YOUR_EMAIL@gmail.com.pickle'),) in /usr/local/lib/python3.12/site-packages/aiohttp/cookiejar.py, line 113: with file_path.open(mode="wb") as f: inside the event loop; This is causing stability issues. Please create a bug report at https://github.com/home-assistant/core/issues?q=is%3Aopen+is%3Aissue For developers, please see https://developers.home-assistant.io/docs/asyncio_blocking_operations/#open

> Detected blocking call to open with args (PosixPath('/config/.storage/alexa_media.YOUR_EMAIL@gmail.com.pickle'),) inside the event loop by custom integration 'alexa_media' at custom_components/alexa_media/__init__.py, line 251: cookies = await login.load_cookie() (offender: /usr/local/lib/python3.12/site-packages/aiohttp/cookiejar.py, line 118: with file_path.open(mode="rb") as f:), please create a bug report at https://github.com/alandtse/alexa_media_player/issues For developers, please see https://developers.home-assistant.io/docs/asyncio_blocking_operations/#open

## Related issue number

Fixes [github.com/alandtse/alexa_media_player Issue #2520](https://github.com/alandtse/alexa_media_player/issues/2520)

## Checklist

- [X] I think the code is well written
- [ ] Unit tests for the changes exist
- [ ] Documentation reflects the changes
- [X] If you provide code modification, please add yourself to `CONTRIBUTORS.txt`
- [X] Add a new news fragment into the `CHANGES/` folder

### CODE REVIEW

File: aiohttp/cookiejar.py [hunk 1]
```diff
@@ -117,16 +121,38 @@ def __init__(
...
         with file_path.open(mode="wb") as f:
             pickle.dump(self._cookies, f, pickle.HIGHEST_PROTOCOL)
+        self._cookies_saved.set()
 
     def load(self, file_path: PathLike) -> None:
+        self._cookies_loaded.clear()
+        loop = asyncio.get_event_loop()
+        loop.run_in_executor(None, self._load_cookies, file_path)
+        # Wait until loading is complete
+        self._cookies_loaded.wait()
```
Comment (bdraco) [inline aiohttp/cookiejar.py:146]: This will block the event loop for the duration of the operation in the thread

**Comment by bdraco** [comment]
I don't think there is a way to make `load` and `save` async safe without making them coroutines

The simplest downstream fix would be to run `load` and `save` in the executor in downstream code

A good change for aiohttp would be to document that the below methods are not async safe and should be run in the executor / another thread.

https://docs.aiohttp.org/en/stable/client_reference.html#aiohttp.CookieJar.load
https://docs.aiohttp.org/en/stable/client_reference.html#aiohttp.CookieJar.save

**Comment by bdraco** [comment]
I aborted the CI run because the tests where hanging for over an hour

File: aiohttp/cookiejar.py (hunk 1 above)
Comment (danielbrunt57) [reply aiohttp/cookiejar.py:146]: So if I understand this correctly then, the event loop is still blocked while the code is now being run in the executor so the net effect is no real change in behaviour except that HA is now unable to detect and warn of a blocking operation in the event loop?

**Comment by danielbrunt57** [comment]
> Thanks for the PR.
> 
> I don't think there is no way to easily make `load` and `save` async safe without making them coroutines
> 
> The simplest downstream fix would be to run `load` and `save` in the executor in downstream code
> 
> A good change for aiohttp would be to document that the below methods are not async safe and should be run in the executor / another thread.
> 
> https://docs.aiohttp.org/en/stable/client_reference.html#aiohttp.CookieJar.load https://docs.aiohttp.org/en/stable/client_reference.html#aiohttp.CookieJar.save

So it looks like I need to figure out how to offload the cookiejar.load/save code in alexapy then...

image

**Comment by danielbrunt57** [comment]
> I aborted the CI run because the tests where hanging for over an hour

I'm still fairly new to the ins-and-outs of PR's and this library has stuff I've not seen before...

**Comment by danielbrunt57** [comment]
This code seems to work for me:
```
                elif isinstance(cookies, defaultdict):
                    _LOGGER.debug("Trying to load aiohttpCookieJar to session")
                    cookie_jar: aiohttp.CookieJar = self._session.cookie_jar
                    loop = asyncio.get_event_loop()
                    try:
                        cookie_jar = await loop.run_in_executor(None, cookie_jar.load, cookiefile)
                        return_cookies = self._get_cookies_from_session()
                        numcookies = len(return_cookies)
```
```
            if cookiefile == self._cookiefile[0]:
                cookie_jar = self._session.cookie_jar
                assert isinstance(cookie_jar, aiohttp.CookieJar)
                if self._debug:
                    _LOGGER.debug("Saving cookie to %s", cookiefile)
                loop = asyncio.get_event_loop()
                try:
                    await loop.run_in_executor(None, cookie_jar.save, cookiefile[0])
```
I'll close this PR and issue a new one for documentation update as you recommended.
Thanks!

File: aiohttp/cookiejar.py (hunk 1 above)
Comment (Dreamsorcerer) [reply aiohttp/cookiejar.py:146]: If there's not await or async for/with, then it's not yielding back to the event loop, so it'd wait here till the function completes.

**Pull request closed without merging**